airbyte-integrations/connectors/source-chargebee/source_chargebee/streams.py

```diff
@@ -0,0 +1,172 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from datetime import datetime, timezone
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional
+
+import backoff
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams import Stream
+from chargebee.api_error import OperationFailedError
+from chargebee.list_result import ListResult
+from chargebee.model import Model
+from chargebee.models import Addon, Customer, Invoice, Order, Plan, Subscription
+
+# Backoff params below
+# according to Chargebee's guidance on rate limit
+# https://apidocs.chargebee.com/docs/api?prod_cat_ver=2#api_rate_limits
+MAX_TRIES = 10  # arbitrary max_tries
+MAX_TIME = 90  # because Chargebee API enforce a per-minute limit
+
+
+class ChargebeeStream(Stream):
+    primary_key = "id"
+    cursor_field = "updated_at"
+    page_size = 100
+    api: Model = None
+
+    def __init__(self, start_date: str):
+        # Convert `start_date` to timestamp(UTC).
+        self._start_date = int(datetime.strptime(start_date, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=timezone.utc).timestamp())
+
+    @property
+    def state_checkpoint_interval(self) -> int:
+        return self.page_size
+
+    def next_page_token(self, list_result: ListResult) -> Optional[Mapping[str, Any]]:
+        # Reference for Chargebee's pagination strategy below:
+        # https://apidocs.chargebee.com/docs/api/#pagination_and_filtering
+        next_offset = list_result.next_offset
+        if next_offset:
+            return {"offset": next_offset}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {
+            "limit": self.page_size,
+            "include_deleted": "true",
+            "sort_by[asc]": self.cursor_field,
+        }
+
+        if next_page_token:
+            params.update(next_page_token)
+
+        start_point = self._start_date
+        if stream_state:
+            start_point = max(start_point, stream_state[self.cursor_field])
+
+        params[f"{self.cursor_field}[after]"] = start_point
+
+        return params
+
+    def parse_response(self, list_result: ListResult, **kwargs) -> Iterable[Mapping]:
+        for message in list_result:
+            yield message._response[self.name]
+
+    @backoff.on_exception(
+        backoff.expo,  # Exponential back-off
+        OperationFailedError,  # Only on Chargebee's OperationFailedError
```

Comment:
Does this mean that we will handle any error as Rate Limit?

Reply:
Done. Added custom backoff handler.